Search.setIndex({"docnames": ["api", "genal", "index", "introduction", "modules"], "filenames": ["api.rst", "genal.rst", "index.rst", "introduction.rst", "modules.rst"], "titles": ["The Geno class", "genal package", "genal: A Python Toolkit for Genetic Risk Scoring and Mendelian Randomization", "Introduction", "genal"], "terms": {"index": 2, "modul": [2, 3, 4], "search": [0, 1, 2, 3], "page": [2, 3], "class": [1, 2, 3], "df": [0, 1], "name": [0, 1, 3, 4], "nonam": [], "chr": [0, 1, 3], "po": [0, 1, 3], "snp": [0, 1, 2], "ea": [0, 1, 3], "nea": [0, 1, 3], "beta": [0, 1, 3], "se": [0, 1, 3], "p": [0, 1, 3], "eaf": [0, 1, 3], "preprocess": [0, 1, 2], "1": [0, 1, 3], "reference_panel": [0, 1, 3, 4], "eur": [0, 1, 3], "none": [0, 1], "effect_column": [0, 1, 3], "keep_column": [0, 1, 3], "keep_multi": [0, 1], "keep_dup": [0, 1, 3], "fill_snpid": [0, 1], "fill_coordin": [0, 1], "base": [0, 1, 2, 3], "object": [0, 1, 3], "A": [0, 1, 3], "handl": [0, 1, 3], "gwa": [0, 1, 2, 3], "deriv": [0, 1, 2, 3], "data": [0, 1, 2, 4], "includ": [0, 1, 2, 3], "rsid": [0, 1, 3], "genom": [0, 1, 2, 3], "posit": [0, 1, 3], "trait": [0, 1, 2, 3], "effect": [0, 1, 3], "allel": [0, 1, 3], "frequenc": [0, 1, 3], "type": [0, 1, 2], "str": [0, 1, 3], "main": [0, 1, 3], "datafram": [0, 1, 3], "contain": [0, 1, 3], "pd": [0, 1, 3], "data_clump": [0, 1], "initi": [0, 1], "after": [0, 1, 3], "method": [0, 1, 3], "execut": [0, 1, 2, 3], "when": [0, 1], "us": [0, 1, 3], "true": [0, 1, 3], "dure": [0, 1, 3], "option": [0, 1, 3], "phenotyp": [0, 1, 3, 4], "tupl": [0, 1], "individu": [0, 1, 3], "level": [0, 1, 3], "string": [0, 1], "repres": [0, 1, 3], "column": [0, 1, 3], "run": [0, 1, 2, 3], "set_phenotyp": [0, 1, 3, 4], "mr_data": [0, 1, 3, 4], "exposur": [0, 1, 3], "outcom": [0, 1, 3], "query_outcom": [0, 1, 3, 4], "ram": [0, 1, 4], "avail": [0, 1, 3], "memori": [0, 1], "int": [0, 1], "cpu": [0, 1, 4], "number": [0, 1, 3], "check": [0, 1, 3, 4], "list": [0, 1, 3], "perform": [0, 1, 3], "refer": [0, 1, 3], "popul": [0, 1, 2, 3], "info": [0, 1, 3], "adjust": [0, 1, 3], "first": [0, 1, 3], "need": [0, 1, 3], "store": [0, 1, 3], "result": [0, 1, 3], "pr": [1, 2, 3, 4], "comput": [0, 1, 3], "polygen": [0, 1, 2], "risk": [0, 1], "score": [0, 1], "assign": [0, 1], "attribut": [0, 1, 3], "association_test": [0, 1, 3, 4], "estim": [0, 1, 3], "standard": [0, 1, 3, 4], "error": [0, 1, 3], "valu": [0, 1, 3], "extract": [1, 2, 3], "from": [0, 1, 2, 3], "mendelian": [0, 1], "random": [0, 1], "between": [0, 1, 2, 3], "presso": [0, 1, 2, 3], "algorithm": [0, 1], "horizont": [0, 1, 2, 3], "pleiotropi": [0, 1, 2, 3], "correct": [0, 1, 3], "one": [0, 1, 3], "build": [0, 1, 3], "anoth": [0, 1, 3], "ivw": [0, 1, 3], "fe": [0, 1], "uwr": [0, 1], "wm": [0, 1], "pen": [0, 1], "simpl": [0, 1, 3], "median": [0, 1, 3], "sign": [0, 1, 3], "egger": [0, 1, 3], "boot": [0, 1], "action": [0, 1, 3], "2": [0, 1, 3], "eaf_threshold": [0, 1], "0": [0, 1, 2, 3], "42": [0, 1], "heterogen": [0, 1, 3], "fals": [0, 1, 3], "nboot": [0, 1], "10000": [0, 1], "penk": [0, 1], "20": [0, 1], "queri": [0, 1], "paramet": [0, 1], "possibl": [0, 1], "invers": [0, 1, 3], "varianc": [0, 1, 3], "weight": [0, 1, 3], "under": [0, 1], "dispers": [0, 1], "fix": [0, 1, 3], "re": [0, 1], "without": [0, 1, 3], "unweight": [0, 1, 3], "regress": [0, 1, 3], "bootstrap": [0, 1, 3], "penalis": [0, 1, 3], "concord": [0, 1, 3], "test": [0, 1, 2], "default": [0, 1, 3], "i": [0, 1, 2, 3], "how": [0, 1, 3], "treat": [0, 1, 3], "palindrom": [0, 1, 3], "harmon": [0, 1, 3], "accept": [0, 1], "doesn": [0, 1], "t": [0, 1, 3], "flip": [0, 1, 3], "them": [0, 1, 3], "assum": [0, 1, 3], "all": [0, 1, 2, 3], "ar": [0, 1, 3], "forward": [0, 1, 3], "strand": [0, 1, 3], "attempt": [0, 1, 3], "conserv": [0, 1, 3], "3": [0, 1, 3], "remov": [0, 1, 3], "veri": [0, 1, 3], "float": [0, 1], "max": [0, 1], "relev": [0, 1, 3], "bool": [0, 1], "If": [0, 1, 3], "cochran": [0, 1, 3], "": [0, 1, 2, 3], "q": [0, 1, 3], "replic": [0, 1], "penalti": [0, 1], "return": [0, 1, 3], "tabl": [0, 1, 3], "n_iter": [0, 1], "outlier_test": [0, 1], "distortion_test": [0, 1], "significance_p": [0, 1], "05": [0, 1, 2], "detect": [0, 1, 2, 3], "treatment": [0, 1], "don": [0, 1], "gener": [0, 1], "step": [0, 1, 3], "improv": [0, 1], "stabil": [0, 1], "identifi": [0, 1, 3], "outlier": [0, 1, 3], "respons": [0, 1], "global": [0, 1, 3], "p_valu": [0, 1, 3], "signific": [0, 1, 3], "distort": [0, 1, 3], "causal": [0, 1, 2], "befor": [0, 1, 3], "statist": [0, 1, 2, 3], "threshold": [0, 1, 3], "both": [0, 1, 3], "identif": [0, 1], "core": [0, 1], "parallel": [0, 1, 3], "follow": [0, 1, 3], "element": [0, 1], "mod_tabl": [0, 1], "origin": [0, 1, 3], "globaltest": [0, 1], "indic": [0, 1, 3], "presenc": [0, 1, 3], "outliertest": [0, 1], "each": [0, 1, 3], "likelihood": [0, 1, 3], "thi": [0, 1, 2, 3], "being": [0, 1], "set": [0, 1, 3], "nan": [0, 1, 3], "distortiontest": [0, 1], "covar": [0, 1, 3], "conduct": [0, 1], "singl": [0, 1, 2, 3], "against": [0, 1], "requir": [0, 1, 3], "function": [1, 2, 3], "The": [1, 2, 3], "also": [1, 3], "expect": [1, 2, 3], "extract_snp": [0, 1, 4], "have": [0, 1, 2, 3], "been": [0, 1, 3], "call": [0, 1, 2, 3], "prior": [0, 1], "covari": [0, 1, 3], "an": [0, 1, 3], "empti": [0, 1], "quantit": [0, 1, 3], "typic": [0, 1], "done": [0, 1, 3], "make": [0, 1, 2, 3], "more": [0, 1, 3], "interpret": [0, 1, 3], "unclump": [], "updat": [0, 1, 3], "kb": [0, 1, 3], "250": [0, 1, 3], "r2": [0, 1, 3], "p1": [0, 1, 3], "5e": [0, 1, 3], "08": [0, 1], "p2": [0, 1], "01": [0, 1], "linkag": [0, 1, 3], "disequilibrium": [0, 1, 3], "process": [0, 1, 3], "plink": [0, 1, 3], "window": [0, 1, 3], "term": [0, 1], "thousand": [0, 1, 3], "abov": [0, 1], "consid": [0, 1], "8": [0, 1, 3], "post": [0, 1], "further": [0, 1], "filter": [0, 1, 3], "won": [0, 1, 3], "sa": [0, 1, 3], "afr": [0, 1, 3], "amr": [0, 1, 3], "altern": [0, 1], "path": [0, 1, 3], "lead": [0, 1], "specif": [0, 1, 3], "bed": [0, 1, 3], "bim": [0, 1, 3], "fam": [0, 1, 3], "panel": [0, 1, 3], "can": [0, 1, 2, 3], "provid": [0, 1, 2, 3], "copi": [0, 1, 3, 4], "creat": [0, 1, 3], "deep": [0, 1], "instanc": [0, 1, 3], "present": [1, 3], "genet": [0, 1, 3], "file": [0, 1, 3], "split": [0, 1, 3], "chromosom": [0, 1, 3], "replac": [0, 1, 3], "For": [0, 1, 3], "ukb_chr": [0, 1], "_file": [0, 1], "output": [0, 1, 3], "tripl": [0, 1, 3], "tmp_genal": [0, 1, 3], "folder": [0, 1], "format": [0, 1, 2, 3], "_extract_allchr": [0, 1], "which": [0, 1, 3], "ukb": [0, 1], "note": [0, 1], "save": [0, 1, 3, 4], "config": [0, 1], "again": [0, 1, 3], "you": [0, 1, 3], "specifi": [0, 1, 3], "want": [0, 1, 3], "same": [0, 1, 3], "get_reference_panel": [0, 1, 4], "retriev": [0, 1], "doe": [0, 1, 2, 3], "try": [0, 1, 3], "argument": [0, 1, 3], "either": [0, 1, 3], "predefin": [0, 1], "necessari": [0, 1, 3], "valid": [0, 1, 3], "rais": [0, 1], "valueerror": [0, 1], "init_attribut": [], "sever": 3, "relat": [], "guess": [], "whether": [1, 3], "alreadi": 3, "tri": [0, 1, 2], "determin": [0, 1, 3], "amount": [0, 1], "mb": [0, 1], "start": [0, 1, 3], "hg19": [0, 1, 3], "end": [0, 1, 3], "hg38": [0, 1, 3], "extraction_fil": [0, 1], "chain_fil": [0, 1], "liftov": [0, 1, 3], "liftover_path": [0, 1, 3], "current": [0, 1], "target": [0, 1, 3], "place": [0, 1], "print": [0, 1, 3], "space": [0, 1], "delimit": [0, 1], "local": [0, 1], "chain": [0, 1, 3], "filenam": [0, 1], "filepath": [0, 1], "extens": [0, 1, 3], "_lift": [], "txt": [0, 1, 3], "directori": [0, 1], "csv": [0, 1, 3], "given": [0, 1, 3], "calcul": [0, 1], "hasn": [0, 1], "5000": [0, 1, 3], "6": [0, 1, 3], "window_snp": [0, 1, 3], "prepar": [0, 1], "exposure_data": [0, 1], "outcome_data": [0, 1], "readi": [0, 1], "h5": [0, 1], "hdf5": [0, 1], "get": [0, 1, 3], "find": [0, 1, 3], "onli": [0, 1, 3], "width": [0, 1, 3], "look": [0, 1, 3], "minimum": [0, 1, 3], "ld": [0, 1], "than": [0, 1, 3], "x": [0, 1, 3], "awai": [0, 1], "fmt": [0, 1], "sep": [0, 1, 3], "header": [0, 1], "futur": [0, 1, 3], "vcf": [0, 1], "gz": [0, 1], "tab": [0, 1], "otherwis": [], "full": 3, "request": [0, 1], "iid": [0, 1, 3], "pheno": [0, 1, 3], "pheno_typ": [0, 1, 3], "alternate_control": [0, 1, 3], "essenti": [0, 1], "row": [0, 1, 3], "least": [0, 1, 3], "id": [0, 1, 3], "These": [0, 1], "should": [0, 1, 3], "correspond": [0, 1, 3], "depend": [0, 1, 3], "variabl": [0, 1], "infer": [0, 1, 2], "binari": [0, 1, 3], "To": [0, 1, 2, 3], "bypass": [0, 1], "quant": [0, 1, 3], "By": [0, 1, 3], "control": [0, 1, 3], "most": [0, 1, 3], "frequent": [0, 1, 3], "case": [0, 1, 3], "sort_group": [0, 1, 4], "lowest_p": [0, 1], "duplic": [0, 1, 3], "combin": [0, 1], "differ": [0, 1, 3], "retain": [0, 1], "lowest": [0, 1], "accordingli": [0, 1], "found": [0, 1, 3], "linreg": [1, 4], "y": 1, "w": 1, "helper": 1, "linear": 1, "mr_egger_regress": [1, 4], "beta_": 1, "se_e": 1, "beta_o": 1, "se_o": 1, "analysi": [0, 1, 2, 3], "see": [1, 3], "mr_egger_regression_bootstrap": [1, 4], "version": [1, 2, 3], "numpi": 1, "arrai": [1, 3], "size": 1, "variant": [1, 2, 3], "boostrap": [1, 3], "iter": [1, 3], "obtain": [1, 3], "two": [1, 3], "dictionari": [0, 1, 3], "intercept": [1, 3], "b": [1, 2, 3], "coeffici": [1, 3], "pval": [1, 3], "nsnp": [1, 3], "dict": [0, 1], "4": [1, 3], "mr_ivw": [1, 4], "oppos": 1, "mr_ivw_r": [1, 4], "q_df": [1, 3], "degre": 1, "freedom": 1, "q_pval": [1, 3], "squar": 1, "wl": 1, "assess": [1, 3], "across": [1, 2], "instrument": [1, 3], "mr_ivw_f": [1, 4], "mr_pen_wm": [1, 4], "http": [1, 3], "arxiv": 1, "org": [1, 3], "ab": 1, "1606": 1, "03729": 1, "constant": 1, "factor": [0, 1], "mr_sign": [1, 4], "consist": [1, 3], "direct": 1, "e": [1, 3], "directon": 1, "assumpt": 1, "kei": [1, 2], "proport": 1, "minu": 1, "5": [1, 2, 3], "multipli": 1, "Not": 1, "applic": 1, "binomi": 1, "distribut": 1, "exactli": 1, "zero": 1, "evalu": 1, "probabl": 1, "observ": 1, "chanc": 1, "alon": 1, "null": 1, "50": [1, 2, 3], "mr_simple_median": [1, 4], "mr_uwr": [1, 4], "mr_weighted_median": [1, 4], "parallel_bootstrap_func": [1, 4], "weighted_median": [1, 4], "b_iv": 1, "weighted_median_bootstrap": [1, 4], "mr_func": [1, 4], "name_exposur": 1, "wrapper": 1, "detail": [1, 3], "regard": [1, 3], "implement": [1, 3], "here": [1, 3], "na": 1, "compil": 1, "apply_action_2": [1, 4], "eaf_": 1, "eaf_o": 1, "align": 1, "outsid": 1, "intermedi": 1, "rang": 1, "flag": 1, "boundari": 1, "delet": [0, 1, 3], "among": 1, "remain": 1, "ones": [1, 3], "check_required_column": [1, 4], "flip_allel": [1, 4], "harmonize_mr": [1, 4], "df_exposur": 1, "df_outcom": 1, "analys": [1, 2, 3], "code": [1, 3], "maxim": 1, "appli": [0, 1, 3], "verifi": 1, "renam": 1, "merg": [1, 3], "classifi": 1, "invert": 1, "switch": 1, "those": 1, "still": 1, "load_outcome_from_filepath": [1, 4], "load": [1, 2], "load_outcome_from_geno_object": [1, 4], "mrpresso_func": [1, 4], "mr_presso": [1, 4], "query_outcome_func": [1, 4], "absent": 1, "getrss_loo": [1, 4], "beta_e_column": 1, "returniv": 1, "getrandomdata": [1, 4], "1000": [0, 1, 3], "biastest": 1, "parallel_rss_loo": [1, 4], "power_eigen": [1, 4], "n": 1, "association_test_func": [1, 4], "covar_list": 1, "data_pheno": 1, "seri": [1, 3], "oper": [1, 2, 3], "preliminari": 1, "prefix": [0, 1], "set_phenotype_func": [1, 4], "continu": 1, "revers": 1, "modifi": [0, 1, 3], "inconsist": 1, "clump_data": [0, 1, 2, 4], "plink19_path": [], "gpf": [], "gibb": [], "pi": [], "falcon": [], "00_gener": [], "softwar": [], "plink1": [], "9": [1, 3], "input": [0, 1, 3], "ani": [0, 1, 3], "create_bedlist": [1, 4], "bedlist": 1, "output_nam": 1, "not_found": 1, "param": 1, "were": [1, 3], "extract_command_parallel": [1, 4], "task_id": 1, "snp_list_path": 1, "task": 1, "extract_snps_from_combined_data": [1, 4], "extract_snps_func": [0, 1, 2, 4], "snp_list": [0, 1], "dataset": [0, 1, 3], "configur": [0, 1], "typeerror": [0, 1], "incorrect": [0, 1, 3], "handle_multiallelic_vari": [1, 4], "merge_command": 1, "multiallel": [0, 1, 3], "merge_extracted_snp": [], "prepare_snp_list": [1, 4], "process_split_data": [], "prs_func": [1, 4], "inform": [0, 1, 3], "equival": 1, "avoid": [0, 1, 3], "repeat": [], "mandatori": 1, "miss": [0, 1, 3], "wa": [1, 3], "success": [1, 3], "report_snps_not_found": [1, 4], "nrow": 1, "report": 1, "setup_path": [], "user": [1, 2, 3], "combine_geno": [1, 4], "g": [1, 2, 3], "arg": 1, "adjust_column_nam": [1, 4], "sure": [0, 1, 3], "other": [1, 3], "keep": [1, 3], "check_allele_column": [1, 4], "allele_col": 1, "upper": [1, 3], "form": [1, 3], "c": [1, 2, 3], "letter": [1, 3], "unless": 1, "check_argu": [1, 4], "pass": [1, 3], "logic": 1, "invalid": [0, 1, 3], "incompat": 1, "check_beta_column": [1, 4], "odd": [0, 1, 3], "ratio": [0, 1, 3], "log": [0, 1, 3], "transform": [0, 1, 3], "check_int_column": [1, 4], "int_col": 1, "int32": 1, "non": [1, 3], "numer": 1, "check_p_column": [1, 4], "inappropri": 1, "check_snp_column": [1, 4], "delete_tmp": [1, 4], "tmp": 1, "fill_coordinates_func": [1, 4], "reference_panel_df": 1, "fill": [0, 1], "fill_ea_nea": [1, 4], "fill_nea": [1, 4], "fill_se_p": [1, 4], "fill_snpids_func": [1, 4], "some": [1, 3], "snpid": 1, "thei": [1, 3], "remove_na": [1, 4], "save_data": [1, 4], "But": [1, 3], "recogn": 1, "lift_coordin": [], "lo": [], "lift_data": [1, 4], "download": [1, 3], "It": [0, 1, 2, 3], "manual": 1, "larg": [0, 1, 3], "suggest": 1, "lift_data_liftov": 1, "overrid": 1, "genal_tools_path": [], "linux": [], "must": 3, "instal": 2, "your": 3, "system": [], "ucsc": 3, "edu": 3, "do": [2, 3], "chang": 3, "inplac": [], "new": [0, 1, 3], "u": [], "we": 3, "genal_tool": [], "post_lift_oper": [1, 4], "like": [1, 3], "prepare_chain_fil": [1, 4], "apply_proxi": [1, 4], "searchspac": [1, 3], "come": [1, 3], "find_proxi": [1, 4], "best": 1, "exist": [1, 3], "suit": 1, "usual": [1, 2, 3], "restrict": 1, "potenti": [1, 3], "much": 1, "faster": [1, 3], "thread": 1, "whole": [1, 3], "apart": 1, "biallel": 1, "query_outcome_proxi": [1, 4], "snps_to_extract": 1, "snps_df": 1, "well": [1, 2, 3], "addit": [1, 3], "choos": [1, 3], "recomput": 1, "check_bfil": [1, 4], "default_config": [1, 4], "get_plink19_path": [1, 4], "plink19": 1, "get_reference_panel_path": [1, 4], "oserror": 1, "issu": 1, "filenotfounderror": 1, "load_reference_panel": [1, 4], "read_config": [1, 4], "set_plink": [1, 3, 4], "set_reference_fold": [1, 4], "allow": 1, "where": [0, 1, 3], "work": [1, 3], "desir": 1, "temporari": 1, "cannot": 1, "messag": 1, "statu": 1, "write_config": [1, 4], "write": 1, "packag": [0, 2, 3, 4], "submodul": 4, "geno": [2, 3, 4], "clump": [2, 4], "mr": [0, 2, 3, 4], "mrpresso": [0, 3, 4], "lift": [0, 2, 4], "mr_tool": 4, "associ": [0, 2, 4], "extract_pr": [0, 4], "geno_tool": 4, "proxi": [0, 3, 4], "tool": [2, 3, 4], "content": 4, "manipul": 0, "through": [0, 3], "its": [0, 1, 2, 3], "genal": [0, 3], "mr_result": [0, 1, 4], "plot": [0, 1], "mr_plot": [0, 1, 4], "intern": [0, 1], "debug": [0, 1], "purpos": [0, 1, 3], "preprocess_data": [0, 1, 3, 4], "clean": [0, 1, 2, 3], "mode": [0, 1], "phi": [0, 1], "exposure_nam": [0, 1, 3], "outcome_nam": [0, 1, 3], "bandwidth": [0, 1], "kernel": [0, 1], "densiti": [0, 1], "displai": [0, 1], "mr_forest": [0, 1, 4], "scatter": [0, 1], "line": [0, 1, 3], "custom": [0, 1, 2, 3], "label": [0, 1], "axi": [0, 1], "plotnin": [0, 1], "ggplot": [0, 1], "ha": [0, 1, 2, 3], "higher": [0, 1, 3], "exclud": [0, 1, 3], "usc": [0, 1], "python": [0, 1, 3], "slower": [0, 1], "nucleotid": [0, 1, 2, 3], "polymorph": [0, 1, 2, 3], "ad": [0, 1, 3], "fill_delet": [0, 1], "ancestri": [0, 1, 3], "classif": [0, 1], "a1": [0, 1], "a2": [0, 1], "OR": [0, 1, 3], "kept": [0, 1], "defer": [0, 1], "decid": [0, 1], "recent": [0, 1], "instead": [0, 1], "recommend": [0, 1, 3], "update_snpid": [0, 1, 4], "mismatch": [0, 1], "notabl": [0, 1], "lose": [0, 1], "due": [0, 1], "merge_command_parallel": [1, 4], "df_subset": 1, "beta_mod": [1, 4], "beta_iv": 1, "se_iv": 1, "point": [1, 3], "bootstrap_mod": [1, 4], "bootstrap_mode_iter": [1, 4], "mr_simple_mod": [1, 4], "mr_weighted_mod": [1, 4], "df_mr_format": [1, 4], "df_mr": 1, "data_origin": 1, "output_path": 1, "extract_snps_from_split_data": [1, 4], "bedlist_path": 1, "lift_coordinates_liftov": [1, 4], "object_id": 1, "chain_path": 1, "lift_coordinates_python": [1, 4], "tmp_id": 1, "create_tmp": [1, 4], "setup_genetic_path": [1, 4], "author": 2, "cyprien": 2, "rivier": 2, "date": 2, "mar": 2, "25": 2, "2024": 2, "design": [2, 3], "easi": [2, 3], "integr": [2, 3], "collect": [2, 3], "facilit": [2, 3], "wide": [2, 3], "studi": [2, 3], "enabl": [2, 3], "clinic": [2, 3], "workflow": [2, 3], "within": [2, 3], "priorit": [2, 3], "friendli": [2, 3], "intuit": [2, 3], "aim": [2, 3], "reduc": [2, 3], "complex": [2, 3], "research": [2, 3], "despit": [2, 3], "focu": [2, 3], "simplic": [2, 3], "sacrific": [2, 3], "depth": 2, "precis": [2, 3], "maintain": [2, 3], "analyt": [2, 3], "rigour": [2, 3], "while": [2, 3], "benefit": [2, 3], "streamlin": [2, 3], "experi": [2, 3], "draw": 2, "concept": 2, "establish": 2, "r": 2, "twosamplemr": 2, "mendelianrandom": 2, "gwasvcf": 2, "adapt": 2, "proven": 2, "methodologi": 2, "environ": 2, "approach": 2, "ensur": [2, 3], "access": [2, 3], "techniqu": 2, "versatil": 2, "scienc": 2, "latest": 2, "releas": 2, "pip": [2, 3], "introduct": 2, "tutori": 2, "hemani": 2, "2018": 2, "platform": 2, "support": 2, "systemat": 2, "human": 2, "phenom": 2, "zheng": 2, "j": 2, "elsworth": 2, "wade": 2, "kh": 2, "baird": 2, "d": 2, "haberland": 2, "v": 2, "laurin": 2, "burgess": 2, "bowden": 2, "langdon": 2, "tan": 2, "vy": 2, "yarmolinski": 2, "shihab": 2, "timpson": 2, "nj": 2, "evan": 2, "dm": 2, "relton": 2, "martin": 2, "rm": 2, "davei": 2, "smith": 2, "gaunt": 2, "tr": 2, "haycock": 2, "pc": 2, "collabor": 2, "elif": 2, "mai": [2, 3], "30": 2, "7": [2, 3], "e34408": 2, "34408": 2, "pmid": 2, "29846171": 2, "verbanck": 2, "widespread": 2, "relationship": 2, "diseas": 2, "mari": 2, "chia": 2, "yen": 2, "chen": 2, "benjamin": 2, "neal": 2, "ron": 2, "natur": [2, 3], "693": 2, "698": 2, "s41588": [2, 3], "018": [2, 3], "0099": 2, "29686387": 2, "lyon": 2, "2020": 2, "effici": 2, "robust": 2, "storag": 2, "summari": [2, 3], "matthew": 2, "shea": 2, "andrew": 2, "ben": 2, "tom": 2, "gibran": 2, "edoardo": 2, "marcora": 2, "biorxiv": 2, "29": 2, "115824v1": 2, "33441155": 2, "short": 3, "easili": 3, "v1": 3, "certain": 3, "yet": 3, "www": 3, "cog": 3, "onc": 3, "go": 3, "systol": 3, "blood": 3, "pressur": 3, "sbp": 3, "investig": 3, "stroke": 3, "uk": 3, "biobank": 3, "our": 3, "appropri": 3, "independ": 3, "serv": 3, "acut": 3, "ischem": 3, "sensit": 3, "previou": 3, "calibr": 3, "show": 3, "util": 3, "In": 3, "pure": 3, "phenoscann": 3, "todo": 3, "publicli": 3, "com": 3, "articl": 3, "0205": 3, "unzip": 3, "panda": 3, "import": 3, "sbp_gwa": 3, "read_csv": 3, "evangelou_30224653_sbp": 3, "head": 3, "markernam": 3, "allele1": 3, "allele2": 3, "freq1": 3, "stderr": 3, "totalsamples": 3, "n_effect": 3, "10": 3, "100000625": 3, "5660": 3, "0523": 3, "0303": 3, "083940": 3, "738170": 3, "736847": 3, "100000645": 3, "7936": 3, "0200": 3, "0372": 3, "591100": 3, "738168": 3, "735018": 3, "100003242": 3, "8831": 3, "1417": 3, "0469": 3, "002526": 3, "733070": 3, "100003304": 3, "9609": 3, "0245": 3, "0838": 3, "769800": 3, "737054": 3, "663809": 3, "100003785": 3, "6406": 3, "0680": 3, "0313": 3, "029870": 3, "738169": 3, "735681": 3, "now": 3, "central": 3, "piec": 3, "made": 3, "take": 3, "describ": 3, "possibli": 3, "inspect": 3, "filler": 3, "expand": 3, "separ": 3, "sbp_geno": 3, "last": 3, "wish": 3, "read": 3, "readm": 3, "particularli": 3, "correctli": 3, "move": 3, "next": 3, "raw": 3, "sometim": 3, "final": 3, "basic": 3, "command": 3, "mean": 3, "chosen": 3, "integ": 3, "charact": 3, "pleas": 3, "api": 3, "197511": 3, "787": 3, "alwai": 3, "rs7899632": 3, "rs61875309": 3, "rs12258651": 3, "rs72828461": 3, "rs1359508": 3, "7088116": 3, "4474": 3, "0114": 3, "0307": 3, "710600": 3, "99997049": 3, "rs10817273": 3, "7088117": 3, "4529": 3, "0151": 3, "623900": 3, "99997707": 3, "rs11794422": 3, "7088118": 3, "4511": 3, "0224": 3, "0306": 3, "463900": 3, "99998403": 3, "rs10981296": 3, "7088119": 3, "8094": 3, "0204": 3, "0389": 3, "599800": 3, "99998646": 3, "rs10981297": 3, "7088120": 3, "9028": 3, "0184": 3, "0517": 3, "722300": 3, "99999468": 3, "rs10981301": 3, "And": 3, "yourself": 3, "time": 3, "european": 3, "select": 3, "involv": 3, "strongli": 3, "interest": 3, "second": 3, "highli": 3, "correl": 3, "close": 3, "sbp_clump": 3, "warn": 3, "760": 3, "top": 3, "1545": 3, "73594": 3, "unit": 3, "pair": 3, "sbp_pr": 3, "divid": 3, "etc": 3, "latter": 3, "pop_chr1": 3, "pop_chr2": 3, "pop_chr": 3, "mani": 3, "happen": 3, "even": 3, "imput": 3, "chr1": 3, "chr2": 3, "chr3": 3, "chr4": 3, "chr5": 3, "chr6": 3, "chr7": 3, "chr8": 3, "chr9": 3, "chr10": 3, "chr11": 3, "chr12": 3, "chr13": 3, "chr14": 3, "chr15": 3, "chr16": 3, "chr17": 3, "chr18": 3, "chr19": 3, "chr20": 3, "chr21": 3, "chr22": 3, "fileset": 3, "4f4ce6a7_allchr": 3, "complet": 3, "786": 3, "874": 3, "759": 3, "49": 3, "126": 3, "about": 3, "half": 3, "salvag": 3, "high": 3, "out": 3, "578": 3, "455": 3, "208": 3, "13": 3, "524": 3, "1330": 3, "1538": 3, "86": 3, "476": 3, "abl": 3, "571": 3, "becaus": 3, "ident": 3, "limit": 3, "gigastrok": 3, "consortium": 3, "s41586": 3, "022": 3, "05165": 3, "stroke_gwa": 3, "gcst90104539_buildgrch37": 3, "tsv": 3, "base_pair_loc": 3, "effect_allele_frequ": 3, "standard_error": 3, "odds_ratio": 3, "ci_low": 3, "ci_upp": 3, "effect_allel": 3, "other_allel": 3, "29439275": 3, "3569": 3, "0030": 3, "0070": 3, "6658": 3, "003005": 3, "989337": 3, "016861": 3, "85928892": 3, "0639": 3, "0152": 3, "0137": 3, "2686": 3, "984915": 3, "958820": 3, "011720": 3, "128341232": 3, "4613": 3, "0025": 3, "0065": 3, "6998": 3, "002503": 3, "989812": 3, "015357": 3, "62707519": 3, "0536": 3, "3177": 3, "015316": 3, "985514": 3, "046019": 3, "80464120": 3, "9789": 3, "0057": 3, "0254": 3, "8223": 3, "005716": 3, "956874": 3, "057052": 3, "stroke_geno": 3, "put": 3, "successfulli": 3, "b352e412": 3, "1541": 3, "stroke_eur": 3, "1312": 3, "023394": 3, "001132": 3, "100": 3, "000807": 3, "021764": 3, "078648": 3, "781986": 3, "022891": 3, "001423": 3, "021525": 3, "001432": 3, "021480": 3, "001364": 3, "373476": 3, "029312": 3, "003063": 3, "001799": 3, "000865": 3, "037777": 3, "030342": 3, "002093": 3, "002758": 3, "000740": 3, "0015": 3, "backward": 3, "complementari": 3, "dna": 3, "fine": 3, "tune": 3, "give": 3, "2583": 3, "740268": 3, "1311": 3, "As": 3, "influenc": 3, "pathwai": 3, "residu": 3, "sum": 3, "pleiotrop": 3, "rerun": 3, "significantli": 3, "alter": 3, "let": 3, "df_pheno": 3, "One": 3, "Then": 3, "advis": 3, "confus": 3, "sbp_adjust": 3, "htn": 3, "At": 3, "model": 3, "ag": 3, "kind": 3, "e415aab3_allchr": 3, "39131": 3, "logist": 3, "38": 3, "37": 3, "howev": 3, "plan": 3, "account": 3, "scroll": 3, "down": 3, "program": 3, "add": 3, "cart": 3, "declar": 3, "profit": 3, "exec": 3}, "objects": {"": [[1, 0, 0, "-", "genal"]], "genal": [[1, 0, 0, "-", "Geno"], [1, 0, 0, "-", "MR"], [1, 0, 0, "-", "MR_tools"], [1, 0, 0, "-", "MRpresso"], [1, 0, 0, "-", "association"], [1, 0, 0, "-", "clump"], [1, 0, 0, "-", "extract_prs"], [1, 0, 0, "-", "geno_tools"], [1, 0, 0, "-", "lift"], [1, 0, 0, "-", "proxy"], [1, 0, 0, "-", "tools"]], "genal.Geno": [[1, 1, 1, "", "Geno"], [0, 2, 1, "id0", "MR"], [0, 3, 1, "", "MR_data"], [0, 2, 1, "", "MR_forest"], [0, 2, 1, "", "MR_plot"], [0, 3, 1, "", "MR_results"], [0, 2, 1, "id1", "MRpresso"], [0, 2, 1, "id2", "association_test"], [0, 3, 1, "", "checks"], [0, 2, 1, "id3", "clump"], [0, 2, 1, "", "copy"], [0, 3, 1, "", "cpus"], [0, 3, 1, "", "data"], [0, 2, 1, "", "extract_snps"], [0, 2, 1, "", "get_reference_panel"], [0, 2, 1, "id4", "lift"], [1, 4, 1, "", "merge_command_parallel"], [0, 3, 1, "", "name"], [0, 3, 1, "", "phenotype"], [0, 2, 1, "id5", "preprocess_data"], [0, 2, 1, "id6", "prs"], [0, 2, 1, "id7", "query_outcome"], [0, 3, 1, "", "ram"], [0, 3, 1, "", "reference_panel"], [0, 2, 1, "", "save"], [0, 2, 1, "id8", "set_phenotype"], [0, 2, 1, "", "sort_group"], [0, 2, 1, "", "standardize"], [0, 2, 1, "", "update_snpids"]], "genal.Geno.Geno": [[1, 2, 1, "id0", "MR"], [1, 3, 1, "", "MR_data"], [1, 2, 1, "", "MR_forest"], [1, 2, 1, "", "MR_plot"], [1, 3, 1, "", "MR_results"], [1, 2, 1, "id1", "MRpresso"], [1, 2, 1, "id2", "association_test"], [1, 3, 1, "", "checks"], [1, 2, 1, "id3", "clump"], [1, 2, 1, "", "copy"], [1, 3, 1, "", "cpus"], [1, 3, 1, "", "data"], [1, 2, 1, "", "extract_snps"], [1, 2, 1, "", "get_reference_panel"], [1, 2, 1, "id4", "lift"], [1, 3, 1, "", "name"], [1, 3, 1, "", "phenotype"], [1, 2, 1, "id5", "preprocess_data"], [1, 2, 1, "id6", "prs"], [1, 2, 1, "id7", "query_outcome"], [1, 3, 1, "", "ram"], [1, 3, 1, "", "reference_panel"], [1, 2, 1, "", "save"], [1, 2, 1, "id8", "set_phenotype"], [1, 2, 1, "", "sort_group"], [1, 2, 1, "", "standardize"], [1, 2, 1, "", "update_snpids"]], "genal.MR": [[1, 4, 1, "", "beta_mode"], [1, 4, 1, "", "bootstrap_mode"], [1, 4, 1, "", "bootstrap_mode_iteration"], [1, 4, 1, "", "linreg"], [1, 4, 1, "", "mr_egger_regression"], [1, 4, 1, "", "mr_egger_regression_bootstrap"], [1, 4, 1, "", "mr_ivw"], [1, 4, 1, "", "mr_ivw_fe"], [1, 4, 1, "", "mr_ivw_re"], [1, 4, 1, "", "mr_pen_wm"], [1, 4, 1, "", "mr_sign"], [1, 4, 1, "", "mr_simple_median"], [1, 4, 1, "", "mr_simple_mode"], [1, 4, 1, "", "mr_uwr"], [1, 4, 1, "", "mr_weighted_median"], [1, 4, 1, "", "mr_weighted_mode"], [1, 4, 1, "", "parallel_bootstrap_func"], [1, 4, 1, "", "weighted_median"], [1, 4, 1, "", "weighted_median_bootstrap"]], "genal.MR_tools": [[1, 4, 1, "", "MR_func"], [1, 4, 1, "", "apply_action_2"], [1, 4, 1, "", "check_required_columns"], [1, 4, 1, "", "df_mr_formatting"], [1, 4, 1, "", "flip_alleles"], [1, 4, 1, "", "harmonize_MR"], [1, 4, 1, "", "load_outcome_from_filepath"], [1, 4, 1, "", "load_outcome_from_geno_object"], [1, 4, 1, "", "mrpresso_func"], [1, 4, 1, "", "query_outcome_func"]], "genal.MRpresso": [[1, 4, 1, "", "getRSS_LOO"], [1, 4, 1, "", "getRandomData"], [1, 4, 1, "", "mr_presso"], [1, 4, 1, "", "parallel_RSS_LOO"], [1, 4, 1, "", "power_eigen"]], "genal.association": [[1, 4, 1, "", "association_test_func"], [1, 4, 1, "", "set_phenotype_func"]], "genal.clump": [[1, 4, 1, "", "clump_data"]], "genal.extract_prs": [[1, 4, 1, "", "create_bedlist"], [1, 4, 1, "", "extract_command_parallel"], [1, 4, 1, "", "extract_snps_from_combined_data"], [1, 4, 1, "", "extract_snps_from_split_data"], [1, 4, 1, "", "extract_snps_func"], [1, 4, 1, "", "handle_multiallelic_variants"], [1, 4, 1, "", "prepare_snp_list"], [1, 4, 1, "", "prs_func"], [1, 4, 1, "", "report_snps_not_found"]], "genal.geno_tools": [[1, 4, 1, "", "Combine_Geno"], [1, 4, 1, "", "adjust_column_names"], [1, 4, 1, "", "check_allele_column"], [1, 4, 1, "", "check_arguments"], [1, 4, 1, "", "check_beta_column"], [1, 4, 1, "", "check_int_column"], [1, 4, 1, "", "check_p_column"], [1, 4, 1, "", "check_snp_column"], [1, 4, 1, "", "fill_coordinates_func"], [1, 4, 1, "", "fill_ea_nea"], [1, 4, 1, "", "fill_nea"], [1, 4, 1, "", "fill_se_p"], [1, 4, 1, "", "fill_snpids_func"], [1, 4, 1, "", "remove_na"], [1, 4, 1, "", "save_data"]], "genal.lift": [[1, 4, 1, "", "lift_coordinates_liftover"], [1, 4, 1, "", "lift_coordinates_python"], [1, 4, 1, "", "lift_data"], [1, 4, 1, "", "post_lift_operations"], [1, 4, 1, "", "prepare_chain_file"]], "genal.proxy": [[1, 4, 1, "", "apply_proxies"], [1, 4, 1, "", "find_proxies"], [1, 4, 1, "", "query_outcome_proxy"]], "genal.tools": [[1, 4, 1, "", "check_bfiles"], [1, 4, 1, "", "create_tmp"], [1, 4, 1, "", "default_config"], [1, 4, 1, "", "delete_tmp"], [1, 4, 1, "", "get_plink19_path"], [1, 4, 1, "", "get_reference_panel_path"], [1, 4, 1, "", "load_reference_panel"], [1, 4, 1, "", "read_config"], [1, 4, 1, "", "set_plink"], [1, 4, 1, "", "set_reference_folder"], [1, 4, 1, "", "setup_genetic_path"], [1, 4, 1, "", "write_config"]]}, "objtypes": {"0": "py:module", "1": "py:class", "2": "py:method", "3": "py:attribute", "4": "py:function"}, "objnames": {"0": ["py", "module", "Python module"], "1": ["py", "class", "Python class"], "2": ["py", "method", "Python method"], "3": ["py", "attribute", "Python attribute"], "4": ["py", "function", "Python function"]}, "titleterms": {"welcom": [], "genal": [1, 2, 4], "": [], "document": [], "indic": 2, "tabl": 2, "packag": 1, "submodul": 1, "geno": [0, 1], "modul": 1, "mr": 1, "mr_tool": 1, "mrpresso": 1, "associ": [1, 3], "clump": [0, 1, 3], "extract_pr": 1, "geno_tool": 1, "lift": [1, 3], "proxi": 1, "tool": 1, "content": [1, 2], "The": 0, "class": 0, "function": 0, "extract": 0, "pr": 0, "A": 2, "python": 2, "toolkit": 2, "genet": 2, "risk": [2, 3], "score": [2, 3], "mendelian": [2, 3], "random": [2, 3], "refer": 2, "introduct": 3, "instal": 3, "tutori": 3, "data": 3, "load": 3, "preprocess": 3, "polygen": 3, "snp": 3, "test": 3}, "envversion": {"sphinx.domains.c": 3, "sphinx.domains.changeset": 1, "sphinx.domains.citation": 1, "sphinx.domains.cpp": 9, "sphinx.domains.index": 1, "sphinx.domains.javascript": 3, "sphinx.domains.math": 2, "sphinx.domains.python": 4, "sphinx.domains.rst": 2, "sphinx.domains.std": 2, "sphinx": 60}, "alltitles": {"The Geno class": [[0, "the-geno-class"]], "Clumping function": [[0, "clumping-function"]], "Extract and PRS functions": [[0, "extract-and-prs-functions"]], "genal package": [[1, "genal-package"]], "Submodules": [[1, "submodules"]], "genal.GENO module": [[1, "module-genal.Geno"]], "genal.MR module": [[1, "module-genal.MR"]], "genal.MR_tools module": [[1, "module-genal.MR_tools"]], "genal.MRpresso module": [[1, "module-genal.MRpresso"]], "genal.association module": [[1, "module-genal.association"]], "genal.clump module": [[1, "module-genal.clump"]], "genal.extract_prs module": [[1, "module-genal.extract_prs"]], "genal.geno_tools module": [[1, "module-genal.geno_tools"]], "genal.lift module": [[1, "module-genal.lift"]], "genal.proxy module": [[1, "module-genal.proxy"]], "genal.tools module": [[1, "module-genal.tools"]], "Module contents": [[1, "module-genal"]], "genal: A Python Toolkit for Genetic Risk Scoring and Mendelian Randomization": [[2, "genal-a-python-toolkit-for-genetic-risk-scoring-and-mendelian-randomization"]], "Contents": [[2, "contents"]], "Indices and tables": [[2, "indices-and-tables"]], "References": [[2, "references"]], "Introduction": [[3, "introduction"]], "Installation": [[3, "installation"]], "Tutorial": [[3, "tutorial"]], "Data loading": [[3, "data-loading"]], "Data preprocessing": [[3, "data-preprocessing"]], "Clumping": [[3, "clumping"]], "Polygenic Risk Scoring": [[3, "polygenic-risk-scoring"]], "Mendelian Randomization": [[3, "mendelian-randomization"]], "SNP-association testing": [[3, "snp-association-testing"]], "Lifting": [[3, "lifting"]], "genal": [[4, "genal"]]}, "indexentries": {"geno (class in genal)": [[0, "genal.Geno"]], "mr() (genal.geno method)": [[0, "genal.Geno.MR"], [0, "id0"]], "mr_data (genal.geno attribute)": [[0, "genal.Geno.MR_data"]], "mr_forest() (genal.geno method)": [[0, "genal.Geno.MR_forest"]], "mr_plot() (genal.geno method)": [[0, "genal.Geno.MR_plot"]], "mr_results (genal.geno attribute)": [[0, "genal.Geno.MR_results"]], "mrpresso() (genal.geno method)": [[0, "genal.Geno.MRpresso"], [0, "id1"]], "association_test() (genal.geno method)": [[0, "genal.Geno.association_test"], [0, "id2"]], "checks (genal.geno attribute)": [[0, "genal.Geno.checks"]], "clump() (genal.geno method)": [[0, "genal.Geno.clump"], [0, "id3"]], "clump_data() (in module genal.clump)": [[0, "genal.clump.clump_data"], [1, "genal.clump.clump_data"]], "copy() (genal.geno method)": [[0, "genal.Geno.copy"]], "cpus (genal.geno attribute)": [[0, "genal.Geno.cpus"]], "data (genal.geno attribute)": [[0, "genal.Geno.data"]], "extract_snps() (genal.geno method)": [[0, "genal.Geno.extract_snps"]], "extract_snps_func() (in module genal.extract_prs)": [[0, "genal.extract_prs.extract_snps_func"], [1, "genal.extract_prs.extract_snps_func"]], "get_reference_panel() (genal.geno method)": [[0, "genal.Geno.get_reference_panel"]], "lift() (genal.geno method)": [[0, "genal.Geno.lift"], [0, "id4"]], "name (genal.geno attribute)": [[0, "genal.Geno.name"]], "phenotype (genal.geno attribute)": [[0, "genal.Geno.phenotype"]], "preprocess_data() (genal.geno method)": [[0, "genal.Geno.preprocess_data"], [0, "id5"]], "prs() (genal.geno method)": [[0, "genal.Geno.prs"], [0, "id6"]], "query_outcome() (genal.geno method)": [[0, "genal.Geno.query_outcome"], [0, "id7"]], "ram (genal.geno attribute)": [[0, "genal.Geno.ram"]], "reference_panel (genal.geno attribute)": [[0, "genal.Geno.reference_panel"]], "save() (genal.geno method)": [[0, "genal.Geno.save"]], "set_phenotype() (genal.geno method)": [[0, "genal.Geno.set_phenotype"], [0, "id8"]], "sort_group() (genal.geno method)": [[0, "genal.Geno.sort_group"]], "standardize() (genal.geno method)": [[0, "genal.Geno.standardize"]], "update_snpids() (genal.geno method)": [[0, "genal.Geno.update_snpids"]], "combine_geno() (in module genal.geno_tools)": [[1, "genal.geno_tools.Combine_Geno"]], "geno (class in genal.geno)": [[1, "genal.Geno.Geno"]], "mr() (genal.geno.geno method)": [[1, "genal.Geno.Geno.MR"], [1, "id0"]], "mr_data (genal.geno.geno attribute)": [[1, "genal.Geno.Geno.MR_data"]], "mr_forest() (genal.geno.geno method)": [[1, "genal.Geno.Geno.MR_forest"]], "mr_func() (in module genal.mr_tools)": [[1, "genal.MR_tools.MR_func"]], "mr_plot() (genal.geno.geno method)": [[1, "genal.Geno.Geno.MR_plot"]], "mr_results (genal.geno.geno attribute)": [[1, "genal.Geno.Geno.MR_results"]], "mrpresso() (genal.geno.geno method)": [[1, "genal.Geno.Geno.MRpresso"], [1, "id1"]], "adjust_column_names() (in module genal.geno_tools)": [[1, "genal.geno_tools.adjust_column_names"]], "apply_action_2() (in module genal.mr_tools)": [[1, "genal.MR_tools.apply_action_2"]], "apply_proxies() (in module genal.proxy)": [[1, "genal.proxy.apply_proxies"]], "association_test() (genal.geno.geno method)": [[1, "genal.Geno.Geno.association_test"], [1, "id2"]], "association_test_func() (in module genal.association)": [[1, "genal.association.association_test_func"]], "beta_mode() (in module genal.mr)": [[1, "genal.MR.beta_mode"]], "bootstrap_mode() (in module genal.mr)": [[1, "genal.MR.bootstrap_mode"]], "bootstrap_mode_iteration() (in module genal.mr)": [[1, "genal.MR.bootstrap_mode_iteration"]], "check_allele_column() (in module genal.geno_tools)": [[1, "genal.geno_tools.check_allele_column"]], "check_arguments() (in module genal.geno_tools)": [[1, "genal.geno_tools.check_arguments"]], "check_beta_column() (in module genal.geno_tools)": [[1, "genal.geno_tools.check_beta_column"]], "check_bfiles() (in module genal.tools)": [[1, "genal.tools.check_bfiles"]], "check_int_column() (in module genal.geno_tools)": [[1, "genal.geno_tools.check_int_column"]], "check_p_column() (in module genal.geno_tools)": [[1, "genal.geno_tools.check_p_column"]], "check_required_columns() (in module genal.mr_tools)": [[1, "genal.MR_tools.check_required_columns"]], "check_snp_column() (in module genal.geno_tools)": [[1, "genal.geno_tools.check_snp_column"]], "checks (genal.geno.geno attribute)": [[1, "genal.Geno.Geno.checks"]], "clump() (genal.geno.geno method)": [[1, "genal.Geno.Geno.clump"], [1, "id3"]], "copy() (genal.geno.geno method)": [[1, "genal.Geno.Geno.copy"]], "cpus (genal.geno.geno attribute)": [[1, "genal.Geno.Geno.cpus"]], "create_bedlist() (in module genal.extract_prs)": [[1, "genal.extract_prs.create_bedlist"]], "create_tmp() (in module genal.tools)": [[1, "genal.tools.create_tmp"]], "data (genal.geno.geno attribute)": [[1, "genal.Geno.Geno.data"]], "default_config() (in module genal.tools)": [[1, "genal.tools.default_config"]], "delete_tmp() (in module genal.tools)": [[1, "genal.tools.delete_tmp"]], "df_mr_formatting() (in module genal.mr_tools)": [[1, "genal.MR_tools.df_mr_formatting"]], "extract_command_parallel() (in module genal.extract_prs)": [[1, "genal.extract_prs.extract_command_parallel"]], "extract_snps() (genal.geno.geno method)": [[1, "genal.Geno.Geno.extract_snps"]], "extract_snps_from_combined_data() (in module genal.extract_prs)": [[1, "genal.extract_prs.extract_snps_from_combined_data"]], "extract_snps_from_split_data() (in module genal.extract_prs)": [[1, "genal.extract_prs.extract_snps_from_split_data"]], "fill_coordinates_func() (in module genal.geno_tools)": [[1, "genal.geno_tools.fill_coordinates_func"]], "fill_ea_nea() (in module genal.geno_tools)": [[1, "genal.geno_tools.fill_ea_nea"]], "fill_nea() (in module genal.geno_tools)": [[1, "genal.geno_tools.fill_nea"]], "fill_se_p() (in module genal.geno_tools)": [[1, "genal.geno_tools.fill_se_p"]], "fill_snpids_func() (in module genal.geno_tools)": [[1, "genal.geno_tools.fill_snpids_func"]], "find_proxies() (in module genal.proxy)": [[1, "genal.proxy.find_proxies"]], "flip_alleles() (in module genal.mr_tools)": [[1, "genal.MR_tools.flip_alleles"]], "genal": [[1, "module-genal"]], "genal.geno": [[1, "module-genal.Geno"]], "genal.mr": [[1, "module-genal.MR"]], "genal.mr_tools": [[1, "module-genal.MR_tools"]], "genal.mrpresso": [[1, "module-genal.MRpresso"]], "genal.association": [[1, "module-genal.association"]], "genal.clump": [[1, "module-genal.clump"]], "genal.extract_prs": [[1, "module-genal.extract_prs"]], "genal.geno_tools": [[1, "module-genal.geno_tools"]], "genal.lift": [[1, "module-genal.lift"]], "genal.proxy": [[1, "module-genal.proxy"]], "genal.tools": [[1, "module-genal.tools"]], "getrss_loo() (in module genal.mrpresso)": [[1, "genal.MRpresso.getRSS_LOO"]], "getrandomdata() (in module genal.mrpresso)": [[1, "genal.MRpresso.getRandomData"]], "get_plink19_path() (in module genal.tools)": [[1, "genal.tools.get_plink19_path"]], "get_reference_panel() (genal.geno.geno method)": [[1, "genal.Geno.Geno.get_reference_panel"]], "get_reference_panel_path() (in module genal.tools)": [[1, "genal.tools.get_reference_panel_path"]], "handle_multiallelic_variants() (in module genal.extract_prs)": [[1, "genal.extract_prs.handle_multiallelic_variants"]], "harmonize_mr() (in module genal.mr_tools)": [[1, "genal.MR_tools.harmonize_MR"]], "lift() (genal.geno.geno method)": [[1, "genal.Geno.Geno.lift"], [1, "id4"]], "lift_coordinates_liftover() (in module genal.lift)": [[1, "genal.lift.lift_coordinates_liftover"]], "lift_coordinates_python() (in module genal.lift)": [[1, "genal.lift.lift_coordinates_python"]], "lift_data() (in module genal.lift)": [[1, "genal.lift.lift_data"]], "linreg() (in module genal.mr)": [[1, "genal.MR.linreg"]], "load_outcome_from_filepath() (in module genal.mr_tools)": [[1, "genal.MR_tools.load_outcome_from_filepath"]], "load_outcome_from_geno_object() (in module genal.mr_tools)": [[1, "genal.MR_tools.load_outcome_from_geno_object"]], "load_reference_panel() (in module genal.tools)": [[1, "genal.tools.load_reference_panel"]], "merge_command_parallel() (in module genal.geno)": [[1, "genal.Geno.merge_command_parallel"]], "module": [[1, "module-genal"], [1, "module-genal.Geno"], [1, "module-genal.MR"], [1, "module-genal.MR_tools"], [1, "module-genal.MRpresso"], [1, "module-genal.association"], [1, "module-genal.clump"], [1, "module-genal.extract_prs"], [1, "module-genal.geno_tools"], [1, "module-genal.lift"], [1, "module-genal.proxy"], [1, "module-genal.tools"]], "mr_egger_regression() (in module genal.mr)": [[1, "genal.MR.mr_egger_regression"]], "mr_egger_regression_bootstrap() (in module genal.mr)": [[1, "genal.MR.mr_egger_regression_bootstrap"]], "mr_ivw() (in module genal.mr)": [[1, "genal.MR.mr_ivw"]], "mr_ivw_fe() (in module genal.mr)": [[1, "genal.MR.mr_ivw_fe"]], "mr_ivw_re() (in module genal.mr)": [[1, "genal.MR.mr_ivw_re"]], "mr_pen_wm() (in module genal.mr)": [[1, "genal.MR.mr_pen_wm"]], "mr_presso() (in module genal.mrpresso)": [[1, "genal.MRpresso.mr_presso"]], "mr_sign() (in module genal.mr)": [[1, "genal.MR.mr_sign"]], "mr_simple_median() (in module genal.mr)": [[1, "genal.MR.mr_simple_median"]], "mr_simple_mode() (in module genal.mr)": [[1, "genal.MR.mr_simple_mode"]], "mr_uwr() (in module genal.mr)": [[1, "genal.MR.mr_uwr"]], "mr_weighted_median() (in module genal.mr)": [[1, "genal.MR.mr_weighted_median"]], "mr_weighted_mode() (in module genal.mr)": [[1, "genal.MR.mr_weighted_mode"]], "mrpresso_func() (in module genal.mr_tools)": [[1, "genal.MR_tools.mrpresso_func"]], "name (genal.geno.geno attribute)": [[1, "genal.Geno.Geno.name"]], "parallel_rss_loo() (in module genal.mrpresso)": [[1, "genal.MRpresso.parallel_RSS_LOO"]], "parallel_bootstrap_func() (in module genal.mr)": [[1, "genal.MR.parallel_bootstrap_func"]], "phenotype (genal.geno.geno attribute)": [[1, "genal.Geno.Geno.phenotype"]], "post_lift_operations() (in module genal.lift)": [[1, "genal.lift.post_lift_operations"]], "power_eigen() (in module genal.mrpresso)": [[1, "genal.MRpresso.power_eigen"]], "prepare_chain_file() (in module genal.lift)": [[1, "genal.lift.prepare_chain_file"]], "prepare_snp_list() (in module genal.extract_prs)": [[1, "genal.extract_prs.prepare_snp_list"]], "preprocess_data() (genal.geno.geno method)": [[1, "genal.Geno.Geno.preprocess_data"], [1, "id5"]], "prs() (genal.geno.geno method)": [[1, "genal.Geno.Geno.prs"], [1, "id6"]], "prs_func() (in module genal.extract_prs)": [[1, "genal.extract_prs.prs_func"]], "query_outcome() (genal.geno.geno method)": [[1, "genal.Geno.Geno.query_outcome"], [1, "id7"]], "query_outcome_func() (in module genal.mr_tools)": [[1, "genal.MR_tools.query_outcome_func"]], "query_outcome_proxy() (in module genal.proxy)": [[1, "genal.proxy.query_outcome_proxy"]], "ram (genal.geno.geno attribute)": [[1, "genal.Geno.Geno.ram"]], "read_config() (in module genal.tools)": [[1, "genal.tools.read_config"]], "reference_panel (genal.geno.geno attribute)": [[1, "genal.Geno.Geno.reference_panel"]], "remove_na() (in module genal.geno_tools)": [[1, "genal.geno_tools.remove_na"]], "report_snps_not_found() (in module genal.extract_prs)": [[1, "genal.extract_prs.report_snps_not_found"]], "save() (genal.geno.geno method)": [[1, "genal.Geno.Geno.save"]], "save_data() (in module genal.geno_tools)": [[1, "genal.geno_tools.save_data"]], "set_phenotype() (genal.geno.geno method)": [[1, "genal.Geno.Geno.set_phenotype"], [1, "id8"]], "set_phenotype_func() (in module genal.association)": [[1, "genal.association.set_phenotype_func"]], "set_plink() (in module genal.tools)": [[1, "genal.tools.set_plink"]], "set_reference_folder() (in module genal.tools)": [[1, "genal.tools.set_reference_folder"]], "setup_genetic_path() (in module genal.tools)": [[1, "genal.tools.setup_genetic_path"]], "sort_group() (genal.geno.geno method)": [[1, "genal.Geno.Geno.sort_group"]], "standardize() (genal.geno.geno method)": [[1, "genal.Geno.Geno.standardize"]], "update_snpids() (genal.geno.geno method)": [[1, "genal.Geno.Geno.update_snpids"]], "weighted_median() (in module genal.mr)": [[1, "genal.MR.weighted_median"]], "weighted_median_bootstrap() (in module genal.mr)": [[1, "genal.MR.weighted_median_bootstrap"]], "write_config() (in module genal.tools)": [[1, "genal.tools.write_config"]]}})